### PR TITLE
port servicegraph processor code to connector

### PR DESCRIPTION
**Pull request by rlankfo**
**Description:** 
This PR ports the deprecated servicegraph processor code to continue support on the connector.

**Link to tracking Issue:** <Issue number if applicable>
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/26091

**Testing:** <Describe what testing was performed and which tests were added.>
`make test`

**Documentation:**
TODO: Documentation should be updated to include the new names of the featuregates which are changed from `processor.servicegraph.*` to `connector.servicegraph.*`

### CODE REVIEW

File: connector/servicegraphconnector/config.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+import (
+	"time"
+)
+
+// Config defines the configuration options for servicegraphconnector.
+type Config struct {
+
+	// MetricsExporter is the name of the metrics exporter to use to ship metrics.
+	MetricsExporter string `mapstructure:"metrics_exporter"`
```
Comment (jpkrohling) [inline connector/servicegraphconnector/config.go:14]: Do we still need this as a connector? Won't the connector get a metrics consumer on its TracesToMetrics constructor?

File: connector/servicegraphconnector/connector.go [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+		reqDurationBounds:                    bounds,
+		keyToMetric:                          make(map[string]metricSeries),
+		shutdownCh:                           make(chan interface{}),
+	}
+}
+
+func (p *serviceGraphConnector) Start(_ context.Context, host component.Host) error {
+	p.store = store.NewStore(p.config.Store.TTL, p.config.Store.MaxItems, p.onComplete, p.onExpire)
+
+	if p.metricsConsumer == nil {
```
Comment (jpkrohling) [inline connector/servicegraphconnector/connector.go:118]: This looks odd: as a traces to metrics connector, I'd expect the metrics consumer to always be set.

File: connector/servicegraphconnector/factory.go [hunk 3]
```diff
@@ -1,13 +1,72 @@
...
+
+func init() {
+	virtualNodeFeatureGate = featuregate.GlobalRegistry().MustRegister(
+		virtualNodeFeatureGateID,
+		featuregate.StageAlpha,
+		featuregate.WithRegisterDescription("When enabled, when the edge expires, connector checks if it has peer attributes(`db.name, net.sock.peer.addr, net.peer.name, rpc.service, http.url, http.target`), and then aggregate the metrics with virtual node."),
+		featuregate.WithRegisterReferenceURL("https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/17196"),
+	)
+	// TODO: Remove this feature gate when the legacy metric names are removed.
+	legacyMetricNamesFeatureGate = featuregate.GlobalRegistry().MustRegister(
```
Comment (jpkrohling) [inline connector/servicegraphconnector/factory.go:37]: Perhaps it shouldn't even exist in the connector?

File: connector/servicegraphconnector/connector.go [hunk 4]
```diff
@@ -0,0 +1,624 @@
...
+		}
+
+		dimensions.CopyTo(dpCalls.Attributes())
+	}
+
+	return nil
+}
+
+func (p *serviceGraphConnector) collectLatencyMetrics(ilm pmetric.ScopeMetrics) error {
+	// TODO: Remove this once legacy metric names are removed
```
Comment (jpkrohling) [inline connector/servicegraphconnector/connector.go:460]: Perhaps we can start the connector without that?

**Review by jpkrohling — COMMENTED** [review]
I have a question about the metrics consumer when creating a new connector, but other than that, LGTM. I verified that the business of this connector matches the one from the processor.

File: connector/servicegraphconnector/config.go (hunk 1 above)
Comment (rlankfo) [reply connector/servicegraphconnector/config.go:14]: We don't need this anymore. Good catch!

File: connector/servicegraphconnector/connector.go (hunk 2 above)
Comment (rlankfo) [reply connector/servicegraphconnector/connector.go:118]: This should always be set -- this is also left over from the processor logic. I removed these nil checks in another place but missed this one.

**Review by rlankfo — COMMENTED** [review]
Thanks for the review @jpkrohling 

The leftover MetricsExporter on the config and the nil checks on the metricsConsumer were leftover behavior from the processor. I removed some of this behavior initially but missed these spots you called out. 

I also removed the legacy metric name feature gate.

File: connector/servicegraphconnector/factory.go [hunk 5]
```diff
@@ -1,13 +1,65 @@
...
+
+var virtualNodeFeatureGate *featuregate.Gate
+
+func init() {
+	virtualNodeFeatureGate = featuregate.GlobalRegistry().MustRegister(
+		virtualNodeFeatureGateID,
+		featuregate.StageAlpha,
+		featuregate.WithRegisterDescription("When enabled, when the edge expires, connector checks if it has peer attributes(`db.name, net.sock.peer.addr, net.peer.name, rpc.service, http.url, http.target`), and then aggregate the metrics with virtual node."),
+		featuregate.WithRegisterReferenceURL("https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/17196"),
+	)
```
Comment (mapno) [inline connector/servicegraphconnector/factory.go:35]: IMO, it's a good opportunity to remove this gate and instead default to a smaller number of attributes or none

**Review by mapno — COMMENTED** [review]
LGTM.

When is the processor removed?

File: connector/servicegraphconnector/config.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+	// - connection_type
+	// The dimensions will be fetched from the span's attributes. Examples of some conventionally used attributes:
+	// https://github.com/open-telemetry/opentelemetry-collector/blob/main/model/semconv/opentelemetry.go.
+	Dimensions []string `mapstructure:"dimensions"`
+
+	// Store contains the config for the in-memory store used to find requests between services by pairing spans.
+	Store StoreConfig `mapstructure:"store"`
+	// CacheLoop is the time to cleans the cache periodically.
+	CacheLoop time.Duration `mapstructure:"cache_loop"`
+	// CacheLoop is the time to expire old entries from the store periodically.
```
Comment (jpkrohling) [inline connector/servicegraphconnector/config.go:30]:
```suggestion
	// StoreExpirationLoop is the time to expire old entries from the store periodically.
```

**Review by jpkrohling — COMMENTED** [review]
Other than the feature gate, this looks good. I have a question about the future of this component though: should we strive to combine this with the span metrics connector? If this is something to consider for the future, do we need this connector here as an intermediary step between the processor and the span metrics connector?

File: connector/servicegraphconnector/factory.go (hunk 5 above)
Comment (rlankfo) [reply connector/servicegraphconnector/factory.go:35]: thanks @mapno that sounds good to me. I will make these changes and request another review.

**Comment by rlankfo** [comment]
> When is the processor removed?

I didn't have a plan to remove the processor yet, but once this is merged I suppose there's no reason not to remove it? cc @jpkrohling

**Comment by rlankfo** [comment]
> I have a question about the future of this component though: should we strive to combine this with the span metrics connector? If this is something to consider for the future, do we need this connector here as an intermediary step between the processor and the span metrics connector?

@jpkrohling I'd like to see this functionality combined with the span metrics connector.  I think this code change probably makes sense still given that the connector already exists and the processor is deprecated. What do you think?

**Comment by jpkrohling** [comment]
> I'd like to see this functionality combined with the span metrics connector. 

Let's get it there then. The current connector is in alpha state, we can still easily deprecate/remove it. I'll open an issue to discuss that with the current code owner (@albertteoh).

**Comment by jpkrohling** [comment]
Opened https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/26648 to discuss this.

**Pull request closed without merging**